workspaces/global-header/plugins/global-header/src/utils/stringUtils.tsx

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import Typography from '@mui/material/Typography';
+
+/**
+ * Highlights the substring that matches the query term.
+ * @param text The full text to render.
+ * @param query The query term to highlight.
+ * @returns JSX.Element with highlighted matching substring.
+ */
+export const highlightMatch = (text: string, query: string) => {
+  if (!query) return <>{text}</>;
+
+  const regex = new RegExp(`(${query})`, 'i');
+  const parts = text.split(regex);
```

Comment:
It looks really nice when it's working!

But this crashes when you enter some regex characters like `[`

Did you pick up that idea from https://github.com/backstage/backstage/blob/master/plugins/search-react/src/components/HighlightedSearchResultText/HighlightedSearchResultText.tsx ?

Can we reuse that component or check why the origin search didn't crash on that character?

Reply:
The original solution doesn't fit our UI design and it uses predefined `preTag` and `postTag` to split text, which won't break the regular expression.
```
text.split(new RegExp(`(${preTag}.+?${postTag})`)); // tags are safe in regExp
```
While my solution uses user input `query` which could include special characters that break a regular expression.

I added some escape to the input text like this for the fix.
```  
const escapeRegex = (input: string) => input.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
const escapedQuery = escapeRegex(query);
```